### PR TITLE
Support drawing popup frame

### PR DESCRIPTION
**Pull request by ath3**
Usability improvement: `popup-frame`.

On some terminals / themes the popups are hard to differentiate from the background.
This makes (for example) LSP information hard to read, and makes the daily work using ssh sessions not as good as it could be.

With this change popups and menus can be configured have a border, and is easy to tell them apart from background with every theme and terminal.

Before:
popup_before

After:
popup_after

### CODE REVIEW

File: book/src/configuration.md [hunk 1]
```diff
@@ -56,6 +56,7 @@ on unix operating systems.
 | `rulers` | List of column positions at which to display the rulers. Can be overridden by language specific `rulers` in `languages.toml` file. | `[]` |
 | `bufferline` | Renders a line at the top of the editor displaying open buffers. Can be `always`, `never` or `multiple` (only shown if more than one buffer is in use) | `never` |
 | `color-modes` | Whether to color the mode indicator with different colors depending on the mode itself | `false` |
+| `popup-frame` | Draw a frame around popups | `false` |
```
Comment (the-mikedavis) [inline book/src/configuration.md:59]: Let's call this "border" instead of "frame"

**Comment by sudormrfbin** [comment]
Drawing borders for popups feel like a good default whereas having a config option for it feels a bit overkill.

**Comment by the-mikedavis** [comment]
I think an option for it seems ok. I prefer the borderless look but I've seen some discussions and matrix messages about wanting a border around popups

File: helix-term/src/ui/lsp.rs [hunk 2]
```diff
@@ -48,8 +48,9 @@ impl SignatureHelp {
 }
 
 impl Component for SignatureHelp {
-    fn render(&mut self, area: Rect, surface: &mut Buffer, cx: &mut Context) {
+    fn render(&mut self, mut area: Rect, surface: &mut Buffer, cx: &mut Context) {
         let margin = Margin::horizontal(1);
+        area.height -= 1;
```
Comment (the-mikedavis) [inline helix-term/src/ui/lsp.rs:53]: Why the change here?

Comment (the-mikedavis) [reply helix-term/src/ui/lsp.rs]: The `clip_bottom` should probably be gated on the new config option as well - there's a noticeable blank line when the option is off

Comment (ath3) [reply helix-term/src/ui/lsp.rs]: I thought the lack of bottom margin was unintentional since other popups have margins around them, and here we only miss it from the bottom.
I changed this now, so without borders if should look as it originally was.

**Comment by zjp-CN** [comment]
I want this to be default. I use `dracula` theme, and in order to distinguish the popup, I have to modify the theme:

image
image


And I haven't seen any popup in nvim without borders. The same codeactionin nvim with [lspsage.nvim](https://github.com/glepnir/lspsaga.nvim) plugin:
image

**Comment by the-mikedavis** [comment]
This change covers popups (like signature-help) but not menus (completion, code-actions). I think an option to enable/disable borders on both would be good

**Comment by ath3** [comment]
> This change covers popups (like signature-help) but not menus (completion, code-actions). I think an option to enable/disable borders on both would be good

It covers now both menus and popups, with settings to enable one of them, both, or none.

**Review by the-mikedavis — DISMISSED** [review]

**Comment by ath3** [comment]
Updated after the changes to popups having scrollbars.
Also fixed the unnecessary leading newlines in popups displayed for the selected menu item.

**Comment by ath3** [comment]
I decided to take out the changes related to selection popup whitespace since it has nothing to do with borders.
I will make a separate pr for that.

**Comment by ath3** [comment]
I think this pr is ready for merge.
The last force-push was only to fix a merge conflict with newer changes on master.

File: helix-term/src/ui/menu.rs [hunk 3]
```diff
@@ -300,14 +303,24 @@ impl<T: Item + 'static> Component for Menu<T> {
...
         surface.clear_with(area, style);
 
+        let border_config = &cx.editor.config().popup_border;
+
+        let render_borders =
+            border_config == &PopupBorderConfig::All || border_config == &PopupBorderConfig::Menu;
+
+        if render_borders {
+            Widget::render(Block::default().borders(Borders::ALL), area, surface);
+            area = area.clip_top(1).clip_bottom(1);
```
Comment (archseer) [inline helix-term/src/ui/menu.rs:321]: area.inner doesn't work with Borders?

Comment (ath3) [reply helix-term/src/ui/menu.rs]: It does, i changed it to use area.inner now.

Comment (ath3) [reply helix-term/src/ui/menu.rs]: I removed the mut from area in function parameter and instead create a new variable "inner" to make it look more consistent with popup render code.
If you prefer the previous version i can revert this change.

**Comment by SoraTenshi** [comment]
Nice job on the PR, has been a staple in my fork for like ever.
Minor bug that i noticed tho: DAP mode (list all variables - Spc g v) does not properly draw the frame.

**Comment by ath3** [comment]
I cleaned up the code, added popup-borders typable command, and fixed the issues with dap variable popup (thanks @SoraTenshi), and with menu only border setting.

**Comment by ath3** [comment]
I think this PR is from my perspective in a good state and unless requested i dont plan to do any changes to it.
To celebrate this PR's birthday, can i get a review please?

**Comment by ath3** [comment]
@the-mikedavis @archseer

File: helix-term/src/ui/menu.rs [hunk 4]
```diff
@@ -362,15 +378,19 @@ impl<T: Item + 'static> Component for Menu<T> {
...
+            if let Some(cursor) = self.cursor {
+                let offset_from_top = cursor - scroll;
+                let left = &mut surface[(area.left(), area.y + offset_from_top as u16)];
+                left.set_style(selected);
+                let right = &mut surface[(
+                    area.right().saturating_sub(1),
+                    area.y + offset_from_top as u16,
+                )];
+                right.set_style(selected);
+            }
```
Comment (dead10ck) [inline helix-term/src/ui/menu.rs:393]: Apologies, since I'm not super familiar with the rendering code, so I could be misunderstanding something, but this looks like if we want to draw borders around popups that we are deciding not to render the cursor? Why is that?

File: helix-term/src/ui/popup.rs [hunk 5]
```diff
@@ -252,13 +255,29 @@ impl<T: Component> Component for Popup<T> {
...
+        }
+
         self.contents.render(inner, surface, cx);
 
         // render scrollbar if contents do not fit
         if self.has_scrollbar {
-            let win_height = inner.height as usize;
-            let len = self.child_size.1 as usize;
+            let win_height = inner.height as usize - 2 * border;
+            let len = self.child_size.1 as usize - 2 * border;
```
Comment (dead10ck) [inline helix-term/src/ui/popup.rs:280]: Should we do a `saturating_sub` here?

File: helix-view/src/editor.rs [hunk 6]
```diff
@@ -950,6 +961,8 @@ pub struct Editor {
     /// field is set and any old requests are automatically
     /// canceled as a result
     pub completion_request_handle: Option<oneshot::Sender<()>>,
+    pub popup_border: bool,
+    pub menu_border: bool,
```
Comment (dead10ck) [inline helix-view/src/editor.rs:965]: Could these just be helper methods instead of new fields on the `Editor` object?

File: helix-term/src/commands/typed.rs [hunk 7]
```diff
@@ -574,6 +574,46 @@ fn set_line_ending(
...
+        cx.editor.popup_border = borders.0;
+        cx.editor.menu_border = borders.1;
+    } else {
+        cx.editor
+            .set_status("Valid options are: none, popup, menu, all");
+    };
+
+    Ok(())
+}
+
```
Comment (dead10ck) [inline helix-term/src/commands/typed.rs:616]: Does this need its own command? Wouldn't just doing `:set popup-border <choice>` work?

**Review by dead10ck — COMMENTED** [review]
Thanks for your patience on this PR! This is something I'd like to use myself; we just only have so much time to review all the PRs.

File: helix-term/src/ui/menu.rs (hunk 4 above)
Comment (ath3) [reply helix-term/src/ui/menu.rs]: This code colors the first and last cell of the selected line in menu.
With borders turned on, the colors would run through the borders.

File: helix-term/src/ui/popup.rs (hunk 5 above)
Comment (ath3) [reply helix-term/src/ui/popup.rs]: Good catch

File: helix-view/src/editor.rs (hunk 6 above)
Comment (ath3) [reply helix-view/src/editor.rs]: Yes, i changed them

File: helix-term/src/commands/typed.rs (hunk 7 above)
Comment (ath3) [reply helix-term/src/commands/typed.rs]: Yeah, i deleted this command now

**Review by dead10ck — APPROVED** [review]
Ok this looks good to me! I tried it out locally too, and it looks good. Thanks for the PR.

**Comment by ath3** [comment]
Bump

File: helix-term/src/ui/popup.rs [hunk 8]
```diff
@@ -252,13 +255,29 @@ impl<T: Component> Component for Popup<T> {
         let background = cx.editor.theme.get("ui.popup");
         surface.clear_with(area, background);
 
-        let inner = area.inner(&self.margin);
+        let render_borders = cx.editor.popup_border();
+
+        let inner = if self
+            .contents
+            .type_name()
+            .starts_with("helix_term::ui::menu::Menu")
```
Comment (the-mikedavis) [inline helix-term/src/ui/popup.rs:263]: This seems very hacky to me but I don't see a good way around it - there isn't a way to compare the type to `Menu<_>` without dealing with strings like this. I think I'm fine with this for now but it's something we should try to eliminate in the future (though it may take a large refactor to the compositor system to do it)

**Review by the-mikedavis — APPROVED** [review]
This looks good to me but I'll leave it up to @archseer to merge

**Review by archseer — APPROVED** [review]

**Comment by nabinkhanal00** [comment]
How to enable popup borders on 23.04?
